text/0152-telemetry-schemas.md

```diff
@@ -0,0 +1,1124 @@
+  </tr>
+  <tr>
+   <td>1.0.x
+   </td>
+   <td>1.0.y
+   </td>
+   <td>yes, for any x and y.
+   </td>
+  </tr>
+  <tr>
+   <td>1.a.x
+   </td>
+   <td>1.b.x
+   </td>
+   <td>if a&lt;b then yes, otherwise no.
+   </td>
+  </tr>
+  <tr>
+   <td>2.0.0
+   </td>
+   <td>1.x.y
+   </td>
+   <td>no
+   </td>
+  </tr>
+</table>
+
+## OTLP Changes
+
+To allow carrying the Schema URL in emitted telemetry it is necessary to add a
+schema_url field to OTLP messages.
+
+We add schema_url fields to the following messages:
+
+```protobuf
+message ResourceSpans {
+  ...
+  // This schema_url applies to the "resource" field and to all spans and span events
+  // in the "instrumentation_library_spans" except the spans and span events which
+  // have a schema_url specified in the nested InstrumentationLibrarySpans message.
+  string schema_url = 3;
+}
+message InstrumentationLibrarySpans {
+  ...
+  // This schema_url applies to all spans in the "spans" field regardless of the
+  // value of the schema_url field in the outer ResourceSpans message.
+  string schema_url = 3;
+}
+
+message ResourceMetrics {
+  ...
+  // This schema_url applies to the "resource" field and to all metrics in the
+  // "instrumentation_library_metrics" except the metrics which have a schema_url
+  // specified in the nested InstrumentationLibraryMetrics message.
+  string schema_url = 3;
+}
+message InstrumentationLibraryMetrics {
+  ...
+  // This schema_url applies to all metrics in the "metrics" field regardless of the
+  // value of the schema_url field in the outer ResourceMetrics message.
+  string schema_url = 3;
+}
+
+message ResourceLogs {
+  ...
+  // This schema_url applies to the "resource" field and to all logs in the
+  // "instrumentation_library_logs" except the logs which have a schema_url
+  // specified in the nested InstrumentationLibraryLogs message.
+  string schema_url = 3;
+}
+message InstrumentationLibraryLogs {
+  ...
+  // This schema_url applies to all logs in the "logs" field regardless of the
+  // value of the schema_url field in the outer ResourceLogs message.
+  string schema_url = 3;
+}
+```
+
+The schema_url field in the ResourceSpans, ResourceMetrics, ResourceLogs
+messages applies to the contained Resource, Span, SpanEvent, Metric, LogRecord
+messages.
+
+The schema_url field in the InstrumentationLibrarySpans message applies to the
+contained Span and SpanEvent messages.
+
+The schema_url field in the InstrumentationLibraryMetrics message applies to the
+contained Metric messages.
+
+The schema_url field in the InstrumentationLibraryLogs message applies to the
+contained LogRecord messages.
+
+If schema_url field is non-empty both in Resource* message and in the contained
+InstrumentationLibrary* message then the value in InstrumentationLibrary*
+message takes the precedence.
+
+## API and SDK Changes
+
+OpenTelemetry API and SDK require the following changes:
+
+- Add method SetSchema(schema_url) to Tracer. After this call all telemetry
```

Comment:
I think it would be better to only allow this to be called once, e.g. by making it an optional argument to the Provider.

Reply:
This is good point and I was thinking about it, but I was not sure we want to alter the signature of an API that is already declared stable (GetTracer). If I am not wrong it would be a backward compatible API change but it will likely be a breaking ABI change (I am forgetting now if we require ABI compatibility for languages like C++).

Alternatively for languages which allow overloading we could introduce a new GetMeter/GetTracer/GetLogEmitter overload which takes 3 parameters (the 3rd parameter being schema_url).

I am open for suggestions.

Reply:
We need to have a design principle that certain API methods are extensible, and use language-specific constructs to achieve that. E.g. in Java the tracer shouldn't be accessed via `provider.getTracer`, but using a builder. In Python it can be getTracer method with kw args.

Reply:
I agree that this should be set only once, preferably at creation time, and once the tracer has been used at all, that it should not be able to have it's schema value changed. 

For instance, I think in Java, we'd probably just add an overload to the TracerProvider.get() method to allow the specification of a schema URL, and under the hood, we'd default it to whatever schema version the semantic convention library aligned with the SDK version used (or something functionally equivalent to this... we might choose to introduce an optional builder for tracers but it would effectively do the same thing).

Reply:
I rewrote "API and SDK Changes" section to reflect this discussion. Please check.

Reply:
@yurishkuro please have a look at the updated "API and SDK Changes" and see if you are OK with it.